python/ray/data/datasource/tensorflow_datasource.py

```diff
@@ -0,0 +1,67 @@
+import logging
+from typing import TYPE_CHECKING, Callable, Iterator, List
+
+from ray.data.block import Block, BlockMetadata, T
+from ray.data.datasource import Datasource, ReadTask
+
+if TYPE_CHECKING:
+    import tensorflow as tf
+
+logger = logging.getLogger(__name__)
+
+
+class SimpleTensorFlowDatasource(Datasource[T]):
+    """A datasource that let's you use TensorFlow datasets with Ray Data.
+
+    .. warning::
+        ``SimpleTensorFlowDataset`` doesn't support parallel reads. You should only use
+        this datasource for small datasets like MNIST or CIFAR.
+
+    Example:
+        >>> def dataset_factory():
+        ...     return tfds.load("cifar10", split=["train"], as_supervised=True)[0]
+        ...
+        >>> dataset = ray.data.read_datasource(
+        ...     SimpleTensorFlowDatasource(),
+        ...     parallelism=1,
+        ...     dataset_factory=dataset_factory
+        ... )
+        >>> features, label = dataset.take(1)[0]
+        >>> features.shape
+        TensorShape([32, 32, 3])
+        >>> label
+        <tf.Tensor: shape=(), dtype=int64, numpy=7>
+    """
+
+    def prepare_read(
+        self,
+        parallelism: int,
+        dataset_factory: Callable[[], "tf.data.Dataset"],
+    ) -> List[ReadTask]:
+        """Return a read task that loads a TensorFlow dataset.
+
+        Arguments:
+            parallelism: This argument isn't used.
+            dataset_factory: A no-argument function that returns the TensorFlow dataset
+                to be read.
+        """
```

Comment:
I can't raise a `ValueError` if the user passes a `tf.data.Dataset`, because Ray can't serialize a `tf.data.Dataset` in the first place.

Reply:
Add a test case for the error that does get raised when trying to pass it in?

Reply:
I don't think we shouldn't test that. 

We'd be testing that Ray Core can't serialize TensorFlow datasets. IMO, that's not a useful test.